tasks/mesos.yml

```diff
@@ -19,62 +19,99 @@
+  register: master_config
 
 - name: Mesos Slave config file
   template: src=conf-mesos-slave.j2 dest=/etc/default/mesos-slave
-  when: mesos_install_mode == "slave" or mesos_install_mode == "master-slave"
-  notify:
-    - Restart mesos-slave
-    - Cleanup meta/slaves/latest
+  when: '{{ "slave" in mesos_install_mode }}'
+  register: slave_config
+
+- name: Check if upstart dropin path exists
+  stat: path=/etc/init/
+  register: upstart_check
+
+- name: Check if systemd sysadmin unit path exists
+  stat: path=/etc/systemd/system/
+  register: systemd_check
 
 - name: Mesos master upstart script
   template: src=init-mesos-master.j2 dest=/etc/init/mesos-master.conf
-  when:
-    - (mesos_install_mode == "master" or mesos_install_mode == "master-slave")
-    - ansible_service_mgr != 'systemd'
-  notify:
-    - Restart mesos-master
+  when: '{{ "master" in mesos_install_mode and upstart_check.stat.exists }}'
+  register: upstartdropinmaster
 
 - name: Mesos slave upstart script
   template: src=init-mesos-slave.j2 dest=/etc/init/mesos-slave.conf
-  when:
-    - (mesos_install_mode == "slave" or mesos_install_mode == "master-slave")
-    - ansible_service_mgr != 'systemd'
-  notify:
-    - Restart mesos-slave
-
-- name: Check if systemd exists
-  stat: path=/lib/systemd/system/
-  register: systemd_check
+  when: '{{ "slave" in mesos_install_mode and upstart_check.stat.exists }}'
+  register: upstartdropinslave
 
 - name: Mesos master systemd script
-  template: src=mesos-master.service.j2 dest=/lib/systemd/system/mesos-master.service
-  when: (mesos_install_mode == "master" or mesos_install_mode == "master-slave") and systemd_check.stat.exists == true
-  notify:
-    - Reload daemon
-    - Restart mesos-master
+  template: src=mesos-master.service.j2 dest=/etc/systemd/system/mesos-master.service
+  when: '{{ "master" in mesos_install_mode and systemd_check.stat.exists }}'
+  register: systemdunitmaster
 
 - name: Mesos slave systemd script
-  template: src=mesos-slave.service.j2 dest=/lib/systemd/system/mesos-slave.service
-  when: (mesos_install_mode == "slave" or mesos_install_mode == "master-slave") and systemd_check.stat.exists == true
-  notify:
-    - Reload daemon
-    - Restart mesos-slave
+  template: src=mesos-slave.service.j2 dest=/etc/systemd/system/mesos-slave.service
+  when: '{{ "slave" in mesos_install_mode and systemd_check.stat.exists }}'
+  register: systemdunitslave
+
+# While waiting for https://github.com/ansible/ansible-modules-core/issues/191
+- name: Reload systemd
+  command: systemctl --system daemon-reload
+  when: '{{ ansible_service_mgr == "systemd" and (systemdunitmaster.changed|default(False) or systemdunitslave.changed|default(False)) }}'
 
 - name: Enable/Disable mesos-master service
-  service: name=mesos-master enabled="{{ (mesos_install_mode == 'slave') | ternary('no', 'yes') }}"
-  when: systemd_check.stat.exists == true
+  service:
+    name: mesos-master
+    enabled: yes
```

Comment:
+1 to use yml-style tasks (and I'ld propose a larger PR refactor to transform whole role), just inconsistent when most tasks still using old style